okhttp/src/main/java/com/squareup/okhttp/internal/OkHttpProtocolNegotiator.java

```diff
@@ -112,13 +114,69 @@ public String getSelectedProtocol(SSLSocket socket) {
     // byte[] getAlpnSelectedProtocol()
     private static final OptionalMethod<Socket> GET_ALPN_SELECTED_PROTOCOL =
         new OptionalMethod<Socket>(byte[].class, "getAlpnSelectedProtocol");
-    // setAlpnSelectedProtocol(byte[])
+    // setAlpnProtocol(byte[])
     private static final OptionalMethod<Socket> SET_ALPN_PROTOCOLS =
         new OptionalMethod<Socket>(null, "setAlpnProtocols", byte[].class);
     // byte[] getNpnSelectedProtocol()
     private static final OptionalMethod<Socket> GET_NPN_SELECTED_PROTOCOL =
         new OptionalMethod<Socket>(byte[].class, "getNpnSelectedProtocol");
-    private static boolean android;
+    // setNpnProtocol(byte[])
+    private static final OptionalMethod<Socket> SET_NPN_PROTOCOLS =
+        new OptionalMethod<Socket>(null, "setNpnProtocols", byte[].class);
+
+    private enum TlsExtensionType {
+      ALPN_AND_NPN,
+      NPN,
+    }
+
+    private static TlsExtensionType tlsExtensionType;
+
+    static {
+      // Decide which TLS Extension (APLN and NPN) we will use, follow the rules:
+      // 1. If Google Play Services Security Provider is installed, use both
+      // 2. If on Android 5.0 or later, use both, else
+      // 3. If on Android 4.1 or later, use NPN, else
+      // 4. Fail.
+      // TODO(madongfly): Logging.
+
+      // Check if Google Play Services Security Provider is installed.
+      try {
+        Class<?> securityClass = Class.forName("java.security.Security");
```

Comment:
yes, it should be always available, I wrongly assumed the Security is a gms core class and wrote such code, then when I copied it from some import statement, I must be distracted to not even notice it is a Java class. -_-b

Thanks for the catch.